<a href="https://colab.research.google.com/github/raymond98tan/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/Raymond_Tan_LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [70]:
# YOUR WORK HERE
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from scipy.stats import ttest_ind

In [71]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-08-11 22:27:24--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip.1’

student.zip.1       100%[===================>]  20.00K  --.-KB/s    in 0.07s   

2020-08-11 22:27:25 (291 KB/s) - ‘student.zip.1’ saved [20478/20478]



In [72]:
!unzip student.zip

Archive:  student.zip
replace student-mat.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

In [ ]:
!head student-mat.csv

school;sex;age;address;famsize;Pstatus;Medu;Fedu;Mjob;Fjob;reason;guardian;traveltime;studytime;failures;schoolsup;famsup;paid;activities;nursery;higher;internet;romantic;famrel;freetime;goout;Dalc;Walc;health;absences;G1;G2;G3
"GP";"F";18;"U";"GT3";"A";4;4;"at_home";"teacher";"course";"mother";2;2;0;"yes";"no";"no";"no";"yes";"yes";"no";"no";4;3;4;1;1;3;6;"5";"6";6
"GP";"F";17;"U";"GT3";"T";1;1;"at_home";"other";"course";"father";1;2;0;"no";"yes";"no";"no";"no";"yes";"yes";"no";5;3;3;1;1;3;4;"5";"5";6
"GP";"F";15;"U";"LE3";"T";1;1;"at_home";"other";"other";"mother";1;2;3;"yes";"no";"yes";"no";"yes";"yes";"yes";"no";4;3;2;2;3;3;10;"7";"8";10
"GP";"F";15;"U";"GT3";"T";4;2;"health";"services";"home";"mother";1;3;0;"no";"yes";"yes";"yes";"yes";"yes";"yes";"yes";3;2;2;1;1;5;2;"15";"14";15
"GP";"F";16;"U";"GT3";"T";3;3;"other";"other";"home";"father";1;2;0;"no";"yes";"yes";"no";"yes";"yes";"no";"no";4;3;2;1;2;5;4;"6";"10";10
"GP";"M";16;"U";"LE3";"T";4;3;"services";"other";"reputation";"mot

In [ ]:
df = pd.read_csv('student-mat.csv', sep=';')

print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


<font color=red>1) Null Hypothesis:
The two categorical variables are independent

2) Alternative Hypothesis:
The two categorical varaibles are dependent</font>

In [56]:
# YOUR WORK HERE
contingency = pd.crosstab(df['health'], df['G3'])

contingency

G3      0   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  19  20
health                                                                        
1        2   1   1   1   0   4   1   7   5   1   5   4   8   1   2   2   2   0
2        7   0   0   3   0   2   3   8   2   2   4   6   2   2   0   2   1   1
3        8   0   2   3   3   8   7  16  13  11   4   5   4   4   1   1   1   0
4        8   0   1   1   3   5   6  11   7   3   5   5   3   3   1   3   1   0
5       13   0   3   7   3  13  11  14  20  14  13   7  16   6   2   4   0   0

In [57]:
observed = pd.crosstab(df['health'], df['G3'], margins=True)

observed

G3       0  4  5   6  7   8   9  10  11  ...  13  14  15  16  17  18  19  20  All
health                                   ...                                     
1        2  1  1   1  0   4   1   7   5  ...   5   4   8   1   2   2   2   0   47
2        7  0  0   3  0   2   3   8   2  ...   4   6   2   2   0   2   1   1   45
3        8  0  2   3  3   8   7  16  13  ...   4   5   4   4   1   1   1   0   91
4        8  0  1   1  3   5   6  11   7  ...   5   5   3   3   1   3   1   0   66
5       13  0  3   7  3  13  11  14  20  ...  13   7  16   6   2   4   0   0  146
All     38  1  7  15  9  32  28  56  47  ...  31  27  33  16   6  12   5   1  395

[6 rows x 19 columns]

In [58]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 69.21650199423279
p value: 0.43611751736038123
dof: 68
expected_frequencies: 
 [[ 4.52151899  0.11898734  0.83291139  1.78481013  1.07088608  3.80759494
   3.33164557  6.66329114  5.59240506  3.68860759  3.68860759  3.21265823
   3.92658228  1.90379747  0.71392405  1.4278481   0.59493671  0.11898734]
 [ 4.32911392  0.11392405  0.79746835  1.70886076  1.02531646  3.64556962
   3.18987342  6.37974684  5.35443038  3.53164557  3.53164557  3.07594937
   3.75949367  1.82278481  0.6835443   1.36708861  0.56962025  0.11392405]
 [ 8.75443038  0.23037975  1.61265823  3.4556962   2.07341772  7.3721519
   6.45063291 12.90126582 10.8278481   7.14177215  7.14177215  6.22025316
   7.60253165  3.68607595  1.38227848  2.76455696  1.15189873  0.23037975]
 [ 6.34936709  0.16708861  1.16962025  2.50632911  1.50379747  5.34683544
   4.67848101  9.35696203  7.85316456  5.17974684  5.17974684  4.51139241
   5.51392405  2.67341772  1.00253165  2.00506329  0.83544304  0.16708861]
 [14.04556962  0.3696202

### Conclusion

<font color=red>Based on a chi2 of 69.2 and p-value of 0.44, we **fail to reject** the null hypothesis that health and G3 score are independent.


---





---


<font color=red>1) Null Hypothesis:
The two categorical variables are independent

2) Alternative Hypothesis:
The two categorical varaibles are dependent</font>

In [ ]:
contingency2 = pd.crosstab(df['paid'], df['failures'])

contingency2

failures    0   1   2   3
paid                     
no        153  37  10  14
yes       159  13   7   2

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(contingency2)

print("chi2:", chi2)
print("p value:", p_value)
print("dof:", dof)
print("expected_frequencies: \n", expected)

chi2: 18.537217658927684
p value: 0.0003407451938069661
dof: 3
expected_frequencies: 
 [[169.03291139  27.08860759   9.21012658   8.66835443]
 [142.96708861  22.91139241   7.78987342   7.33164557]]


### Conclusion

<font color=red>Based on a chi2 of 18.53 and p-value of 0.00034, we **reject** the null hypothesis that paid and failures are independent.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# YOUR WORK HERE
observed

G3       0  4  5   6  7   8   9  10  11  ...  13  14  15  16  17  18  19  20  All
health                                   ...                                     
1        2  1  1   1  0   4   1   7   5  ...   5   4   8   1   2   2   2   0   47
2        7  0  0   3  0   2   3   8   2  ...   4   6   2   2   0   2   1   1   45
3        8  0  2   3  3   8   7  16  13  ...   4   5   4   4   1   1   1   0   91
4        8  0  1   1  3   5   6  11   7  ...   5   5   3   3   1   3   1   0   66
5       13  0  3   7  3  13  11  14  20  ...  13   7  16   6   2   4   0   0  146
All     38  1  7  15  9  32  28  56  47  ...  31  27  33  16   6  12   5   1  395

[6 rows x 19 columns]

In [64]:
row_totals = df['health'].value_counts().sort_index().values

row_totals

array([ 47,  45,  91,  66, 146])

In [65]:
column_totals = df['G3'].value_counts().sort_index().values

column_totals

array([38,  1,  7, 15,  9, 32, 28, 56, 47, 31, 31, 27, 33, 16,  6, 12,  5,
        1])

In [53]:
sample_size = df.shape[0]

sample_size

395

In [66]:
expected = np.array([[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], 
                     [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], 
                     [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], 
                     [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], 
                     [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]])

for row_index, row in enumerate(row_totals):
  for column_index, col in enumerate(column_totals):
    # calculate the expected value
    expected_value = (row*col)/sample_size
    # put that value into the expected value table
    # at the correct location.
    expected[row_index][column_index] = expected_value
    # print(expected_value, row_index, column_index)

expected

array([[ 4.52151899,  0.11898734,  0.83291139,  1.78481013,  1.07088608,
         3.80759494,  3.33164557,  6.66329114,  5.59240506,  3.68860759,
         3.68860759,  3.21265823,  3.92658228,  1.90379747,  0.71392405,
         1.4278481 ,  0.59493671,  0.11898734],
       [ 4.32911392,  0.11392405,  0.79746835,  1.70886076,  1.02531646,
         3.64556962,  3.18987342,  6.37974684,  5.35443038,  3.53164557,
         3.53164557,  3.07594937,  3.75949367,  1.82278481,  0.6835443 ,
         1.36708861,  0.56962025,  0.11392405],
       [ 8.75443038,  0.23037975,  1.61265823,  3.4556962 ,  2.07341772,
         7.3721519 ,  6.45063291, 12.90126582, 10.8278481 ,  7.14177215,
         7.14177215,  6.22025316,  7.60253165,  3.68607595,  1.38227848,
         2.76455696,  1.15189873,  0.23037975],
       [ 6.34936709,  0.16708861,  1.16962025,  2.50632911,  1.50379747,
         5.34683544,  4.67848101,  9.35696203,  7.85316456,  5.17974684,
         5.17974684,  4.51139241,  5.51392405,  2.673

In [67]:
observed_vals = contingency.values

In [68]:
# Chi2 Statistic

chi2 = ((observed_vals - expected)**2 / expected).sum()

chi2

69.21650199423279

In [69]:
# Degrees of Freedom
(18-1) * (5-1)

68

#### P-Value
<font color=red>Plugging in 68 for Degrees of Freedom and 69.2165 for the chi2 statistic on https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html, we get a p-value of **0.436**</font>

## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

<font color=red> A sample estimate is a statistic that estimates some fact about the population.</font>

- What are hypothesis test? How are they useful?

<font color=red> Hypothesis tests are tests that check assumptions(null/alternative hypotheses) that are made about data. They are useful because we can see how significant two datasets are with each other.</font>

- What is a "null hypothesis?"

<font color=red> The null hypothesis is the hyopthesis that there is no difference between the datasets and we base our conclusions on the results to see whether we are able to reject it or not.</font>
- <font color=red> For 1 sample: sample and population means are equal</font>
- <font color=red>For 2 sample: the two groups are equal</font>
- <font color=red> For Chi2: No relationship exists beteween categories(inpedendent)</font>

- What is a p-value and what does it represent?

<font color=red> To know if a claim is valid or not, we use a p-value to weigh the strength of the evidence to see if it’s statistically significant. The p-value is the likelihood of the sample data falling within the range of the population data. </font>

- What does it mean for something to be "statistically significant?"

<font color=red>If something is statistically significant, it means that the likelihood that a relationship between two or more variables is caused by something other than chance.</font>

- What is a test statistic and how does it relate to a p-value?

<font color=red> The test statistic tells us how many deviations away the sample falls from the population and it relates to the p-value in that as the test statistic increases, the p-value decreases.</font>

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

<font color=red>Degrees of freedom are the number of "observations" in the data that are free to vary when estimating statistical parameters.
- In 1-sample T-test: DoF = n-1 
- In 2-sample T-test: DoF = the smaller of $n_1$ – 1 and $n_2$ – 1
- In 1-sample Chi^2: DoF = # of categories - 1
- In 2-sample Chi^2: DoF = (# rows_crosstab-1)*(# cols_crosstab-1)</font>

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)